### PR TITLE
Add support for certificate authentication

### PR DESCRIPTION
**Pull request by yusmenrashidov**

### CODE REVIEW

File: src/main/java/com/sap/cloud/alert/notification/client/internal/PemCertificateUtils.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.Security;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+public class PemCertificateUtils {
```
Comment (mkunev) [inline src/main/java/com/sap/cloud/alert/notification/client/internal/PemCertificateUtils.java:32]: Why do we need this to be public

File: src/main/java/com/sap/cloud/alert/notification/client/internal/AbstractClient.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+import static org.apache.http.HttpHeaders.*;
+import static org.apache.http.HttpStatus.*;
+
+public class AbstractClient {
+
+    private static final String FAILED_EXECUTION = "Failed execution";
+
+    private static final long TIME_DELTA_IN_MILLIS = 300000;
+
+    long credentialsLoadTime;
```
Comment (mkunev) [inline src/main/java/com/sap/cloud/alert/notification/client/internal/AbstractClient.java:37]: Why are these not private?

File: src/main/java/com/sap/cloud/alert/notification/client/internal/KeyStoreDetails.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.sap.cloud.alert.notification.client.internal;
+
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+@ToString(doNotUseGetters = true)
+@EqualsAndHashCode(doNotUseGetters = true)
+public class KeyStoreDetails {
```
Comment (mkunev) [inline src/main/java/com/sap/cloud/alert/notification/client/internal/KeyStoreDetails.java:8]: Can this be package private?

File: src/main/java/com/sap/cloud/alert/notification/client/internal/KeyStoreType.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.sap.cloud.alert.notification.client.internal;
+
+public enum KeyStoreType {
```
Comment (mkunev) [inline src/main/java/com/sap/cloud/alert/notification/client/internal/KeyStoreType.java:3]: Can this be package private?

File: src/main/java/com/sap/cloud/alert/notification/client/internal/HttpClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.http.conn.ssl.DefaultHostnameVerifier;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.ssl.SSLContexts;
+
+import javax.net.ssl.SSLContext;
+import java.security.KeyStore;
+
+import static com.sap.cloud.alert.notification.client.internal.KeyStoreUtils.buildKeyStore;
+
+public class HttpClientFactory {
```
Comment (mkunev) [inline src/main/java/com/sap/cloud/alert/notification/client/internal/HttpClientFactory.java:14]: package private?

File: src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+    private DestinationContext buildDestinationContext(HttpEntity response) throws IOException {
+        try (InputStream responseBodyAsStream = response.getContent()) {
+            JsonNode responseBody = JSON_OBJECT_MAPPER.readTree(responseBodyAsStream);
+
+            JsonNode destinationConfiguration = responseBody.get(DESTINATION_CONFIGURATION_KEY);
+
+            String serviceUri = destinationConfiguration.get(DESTINATION_URL_KEY).asText();
+            String authenticationType = destinationConfiguration.get(AUTHENTICATION_KEY).asText();
+            assertValidAuthenticationType(authenticationType);
```
Comment (mkunev) [inline src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java:122]: Can we have OAuth or Basic here?

File: src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    private IAuthorizationHeader createAuthorizationHeader(HttpEntity response) throws IOException {
+        try (InputStream responseBodyAsStream = response.getContent()) {
+            JsonNode responseBody = JSON_OBJECT_MAPPER.readTree(responseBodyAsStream);
+
+            JsonNode destinationConfiguration = responseBody.get(DESTINATION_CONFIGURATION_KEY);
+            String authenticationType = destinationConfiguration.get(AUTHENTICATION_KEY).asText();
+
+            assertValidAuthenticationType(authenticationType);
```
Comment (mkunev) [inline src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java:136]: Can we have Certificate service here?

File: src/main/java/com/sap/cloud/alert/notification/client/internal/HttpClientFactory.java (hunk 5 above)
Comment (yusmenrashidov) [reply src/main/java/com/sap/cloud/alert/notification/client/internal/HttpClientFactory.java:14]: it is used in builders, that are outside internal package

File: src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java (hunk 6 above)
Comment (yusmenrashidov) [reply src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java:122]: Sure, basic and oauth are also supported

File: src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java (hunk 7 above)
Comment (yusmenrashidov) [reply src/main/java/com/sap/cloud/alert/notification/client/internal/DestinationCredentialsProvider.java:136]: If you mean support from Destination service side, there is not a specific type for Certificate service

File: src/main/java/com/sap/cloud/alert/notification/client/internal/AbstractClient.java (hunk 2 above)
Comment (yusmenrashidov) [reply src/main/java/com/sap/cloud/alert/notification/client/internal/AbstractClient.java:37]: Value is modified for in tests

File: src/main/java/com/sap/cloud/alert/notification/client/internal/KeyStoreDetails.java (hunk 3 above)
Comment (yusmenrashidov) [reply src/main/java/com/sap/cloud/alert/notification/client/internal/KeyStoreDetails.java:8]: This class is also used in builders, which are located in different package

**Review by ipaunov — APPROVED** [review]

File: src/main/java/com/sap/cloud/alert/notification/client/builder/AlertNotificationClientBuilder.java [hunk 8]
```diff
@@ -1,23 +1,32 @@
...
 
 import static com.sap.cloud.alert.notification.client.Platform.CF;
-import static java.util.Objects.isNull;
-import static java.util.Objects.requireNonNull;
+import static java.util.Objects.*;
 
 public class AlertNotificationClientBuilder {
 
+    private String username;
+    private String password;
```
Comment (sbabev) [inline src/main/java/com/sap/cloud/alert/notification/client/builder/AlertNotificationClientBuilder.java:20]: Can't we rename username & password to clientId, clientSecret?

**Review by sbabev — APPROVED** [review]

**Review by ipaunov — APPROVED** [review]

**Review by ipaunov — APPROVED** [review]

**Review by sbabev — APPROVED** [review]

**Pull request merged**